File: Sistema de Calidad/app/src/main/java/com/example/sistemadecalidad/data/model/Fichado.kt
package com.example.sistemadecalidad.data.model

import com.google.gson.annotations.SerializedName

/**
 * Modelo de datos para el fichado/asistencia
 * Basado en la estructura de la tabla asistencia del backend
 */
data class Fichado(
    @SerializedName("id")
    val id: Int,
    
    @SerializedName("usuario_id")
    val usuarioId: Int,
    
    @SerializedName("fecha")
    val fecha: String,
    
    @SerializedName("hora_entrada")
    val horaEntrada: String?,
    
    @SerializedName("hora_salida")
    val horaSalida: String?,
    
    @SerializedName("latitud")
    val latitud: Double?,
    
    @SerializedName("longitud")
    val longitud: Double?,
    
    @SerializedName("ubicacion_valida")
    val ubicacionValida: Boolean?,
    
    @SerializedName("codigo_qr")
    val codigoQr: String?,
    
    @SerializedName("metodo_fichado")
    val metodoFichado: String, // "MANUAL", "GPS", "QR"
    
    @SerializedName("observaciones")
    val observaciones: String?,
    
    @SerializedName("timestamp_creacion")
    val timestampCreacion: String?
)

/**
 * Modelo para la petición de fichado de entrada
 */
data class FichadoEntradaRequest(
    @SerializedName("metodo")
    val metodo: String = "MANUAL",
    
    @SerializedName("latitud")
    val latitud: Double? = null,
    
    @SerializedName("longitud")
    val longitud: Double? = null,
    
    @SerializedName("codigo_qr")
    val codigoQr: String? = null
)

/**
 * Modelo para la petición de fichado de salida
 */
data class FichadoSalidaRequest(
    @SerializedName("metodo")
    val metodo: String = "MANUAL",
    
    @SerializedName("latitud")
    val latitud: Double? = null,
    
    @SerializedName("longitud")
    val longitud: Double? = null
)

/**
 * Modelo para la respuesta de fichado
 */
data class FichadoResponse(
    @SerializedName("success")
    val success: Boolean,
    
    @SerializedName("message")
    val message: String?,
    
    @SerializedName("hora")
    val hora: String?,
    
    @SerializedName("qr_valido")
    val qrValido: Boolean?,
    
    @SerializedName("ubicacion_valida")
    val ubicacionValida: Boolean?,
    
    @SerializedName("distancia_metros")
    val distanciaMetros: Double?,
    
    @SerializedName("horas_trabajadas")
    val horasTrabajadas: String?,
    
    @SerializedName("error")
    val error: String?
)

/**
 * Modelo para el historial de fichados
 */
data class HistorialResponse(
    @SerializedName("success")
    val success: Boolean,
    
    @SerializedName("data")
    val data: List<FichadoHistorial>?,
    
    @SerializedName("error")
    val error: String?
)

/**
 * Modelo de fichado para el historial (estructura del backend)
 */
data class FichadoHistorial(
    @SerializedName("id")
    val id: Int,
    
    @SerializedName("fecha")
    val fecha: String,
    
    @SerializedName("hora_entrada")
    val horaEntrada: String?,
    
    @SerializedName("hora_salida")
    val horaSalida: String?,
    
    @SerializedName("horas_trabajadas")
    val horasTrabajadas: Double?,
    
    @SerializedName("latitud")
    val latitud: Double?,
    
    @SerializedName("longitud")
    val longitud: Double?,
    
    @SerializedName("latitud_salida")
    val latitudSalida: Double?,
    
    @SerializedName("longitud_salida")
    val longitudSalida: Double?,
    
    @SerializedName("metodo_fichado")
    val metodoFichado: String,
    
    @SerializedName("observaciones")
    val observaciones: String?,
    
    @SerializedName("timestamp_creacion")
    val timestampCreacion: String?,
    
    @SerializedName("estado_gps")
    val estadoGps: String?
)

/**
 * Modelo para el dashboard del día actual
 */
data class DashboardHoyResponse(
    @SerializedName("success")
    val success: Boolean,
    
    @SerializedName("data")
    val data: DashboardData?,
    
    @SerializedName("error")
    val error: String?
)

/**
 * Modelo de datos del dashboard
 */
data class DashboardData(
    @SerializedName("timestamp")
    val timestamp: String?,
    
    @SerializedName("fecha_actual")
    val fechaActual: String?,
    
    @SerializedName("hora_actual")
    val horaActual: String?,
    
    @SerializedName("fecha_completa")
    val fechaCompleta: String?,
    
    @SerializedName("dia_semana")
    val diaSemana: String?,
    
    @SerializedName("usuario")
    val usuario: UsuarioDashboard?,
    
    @SerializedName("estado_fichado")
    val estadoFichado: EstadoFichado?,
    
    @SerializedName("server_time")
    val serverTime: String?
)

/**
 * Usuario en dashboard
 */
data class UsuarioDashboard(
    @SerializedName("id")
    val id: Int,
    
    @SerializedName("nombre")
    val nombre: String,
    
    @SerializedName("email")
    val email: String,
    
    @SerializedName("rol")
    val rol: String
)

/**
 * Estado del fichado actual
 */
data class EstadoFichado(
    @SerializedName("tiene_entrada")
    val tieneEntrada: Boolean,
    
    @SerializedName("tiene_salida")
    val tieneSalida: Boolean,
    
    @SerializedName("puede_marcar_entrada")
    val puedemarcarEntrada: Boolean,
    
    @SerializedName("puede_marcar_salida")
    val puedeMarcarSalida: Boolean,
    
    @SerializedName("hora_entrada")
    val horaEntrada: String?,
    
    @SerializedName("hora_salida")
    val horaSalida: String?,
    
    @SerializedName("horas_trabajadas")
    val horasTrabajadas: Double?,
    
    @SerializedName("tiempo_transcurrido")
    val tiempoTranscurrido: Double?,
    
    @SerializedName("metodo")
    val metodo: String?,
    
    @SerializedName("observaciones")
    val observaciones: String?,
    
    @SerializedName("timestamp_creacion")
    val timestampCreacion: String?,
    
    @SerializedName("gps_info")
    val gpsInfo: GpsInfo?
)

/**
 * Información GPS del fichado
 */
data class GpsInfo(
    @SerializedName("entrada_con_gps")
    val entradaConGps: Boolean,
    
    @SerializedName("salida_con_gps")
    val salidaConGps: Boolean,
    
    @SerializedName("coordenadas_entrada")
    val coordenadasEntrada: Coordenadas?,
    
    @SerializedName("coordenadas_salida")
    val coordenadasSalida: Coordenadas?
)

/**
 * Coordenadas GPS
 */
data class Coordenadas(
    @SerializedName("latitud")
    val latitud: Double,
    
    @SerializedName("longitud")
    val longitud: Double
)

/**
 * Modelo para el resumen de analítica
 */
data class ResumenResponse(
    @SerializedName("success")
    val success: Boolean,
    
    @SerializedName("horas_semana")
    val horasSemana: String?,
    
    @SerializedName("horas_mes")
    val horasMes: String?,
    
    @SerializedName("promedio_entrada")
    val promedioEntrada: String?,
    
    @SerializedName("promedio_salida")
    val promedioSalida: String?,
    
    @SerializedName("dias_trabajados_mes")
    val diasTrabajadosMes: Int?,
    
    @SerializedName("error")
    val error: String?
)

File: Backend/routes/fichado.js
const express = require('express');
const { db } = require('../utils/database');
const { authenticateToken } = require('../middleware/auth');
const { requireGPSValidation, validateGPSLocation } = require('../utils/gpsValidation');
const { formatDateForDB, formatTimeForDB } = require('../utils/timeUtils');

const router = express.Router();

// POST /api/fichado/entrada - Registrar entrada con validación GPS
router.post('/entrada', authenticateToken, requireGPSValidation(true), (req, res) => {
    try {
        const { metodo = 'GPS', latitud, longitud, codigo_qr, observaciones } = req.body;
        const usuarioId = req.user.id;
        
        // Usar zona horaria de Perú para consistencia
        const fecha = formatDateForDB();
        const horaEntrada = formatTimeForDB();

        // Verificar si ya tiene entrada SIN SALIDA (no completada)
        db.get(
            'SELECT id FROM asistencia WHERE usuario_id = ? AND fecha = ? AND hora_entrada IS NOT NULL AND hora_salida IS NULL',
            [usuarioId, fecha],
            (err, entradaAbierta) => {
                if (err) {
                    console.error('Error verificando entrada abierta:', err);
                    return res.status(500).json({
                        success: false,
                        error: 'Error interno del servidor'
                    });
                }

                if (entradaAbierta) {
                    console.log(`⚠️ Ya existe entrada sin salida para usuario ${usuarioId} en fecha ${fecha}`);
                    return res.status(400).json({
                        success: false,
                        error: 'Entrada pendiente',
                        message: 'Debes marcar salida antes de registrar nueva entrada'
                    });
                }

                // Registrar nueva entrada - PERMITIR MÚLTIPLES ENTRADAS/SALIDAS EN EL MISMO DÍA
                console.log(`📝 Registrando nueva entrada: Usuario ${usuarioId}, Fecha ${fecha}, Hora ${horaEntrada}`);
                const insertQuery = `
                    INSERT INTO asistencia 
                    (usuario_id, fecha, hora_entrada, latitud, longitud, codigo_qr, metodo_fichado, observaciones)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                `;

                const observacionesCompletas = [
                    observaciones,
                    req.gpsValidation ? `GPS validado - Distancia: ${req.gpsValidation.distance}m` : null
                ].filter(Boolean).join(' | ');

                console.log(`📌 Datos a insertar:`, { usuarioId, fecha, horaEntrada, latitud, longitud, metodo, observacionesCompletas });

                db.run(
                    insertQuery,
                    [usuarioId, fecha, horaEntrada, latitud, longitud, codigo_qr, metodo, observacionesCompletas],
                    function(err) {
                        if (err) {
                            console.error('❌ Error registrando entrada:', err);
                            return res.status(500).json({
                                success: false,
                                error: 'Error registrando entrada'
                            });
                        }

                        console.log(`✅ Entrada registrada exitosamente! ID: ${this.lastID}`);
                        res.json({
                            success: true,
                            message: 'Entrada registrada correctamente',
                            data: {
                                id: this.lastID,
                                fecha,
                                hora_entrada: horaEntrada,
                                metodo,
                                usuario: req.user.nombre,
                                gps_validation: req.gpsValidation ? {
                                    distance: req.gpsValidation.distance,
                                    max_distance: req.gpsValidation.maxDistance,
                                    location_valid: req.gpsValidation.isValid
                                } : null
                            }
                        });
                    }
                );
            }
        );

    } catch (error) {
        console.error('Error en fichado entrada:', error);
        res.status(500).json({
            success: false,
            error: 'Error interno del servidor'
        });
    }
});

// POST /api/fichado/salida - Registrar salida con validación GPS opcional
router.post('/salida', authenticateToken, requireGPSValidation(false), (req, res) => {
    try {
        const { latitud, longitud, observaciones } = req.body;
        const usuarioId = req.user.id;
        
        // Usar zona horaria de Perú para consistencia
        const fecha = formatDateForDB();
        const horaSalida = formatTimeForDB();

        // Buscar la ÚLTIMA entrada sin salida del día (para soportar múltiples turnos)
        db.get(
            'SELECT * FROM asistencia WHERE usuario_id = ? AND fecha = ? AND hora_entrada IS NOT NULL AND hora_salida IS NULL ORDER BY hora_entrada DESC LIMIT 1',
            [usuarioId, fecha],
            (err, entry) => {
                if (err) {
                    console.error('Error buscando entrada:', err);
                    return res.status(500).json({
                        success: false,
                        error: 'Error interno del servidor'
                    });
                }

                if (!entry) {
                    return res.status(400).json({
                        success: false,
                        error: 'Sin entrada registrada',
                        message: 'No tienes una entrada sin salida para hoy'
                    });
                }

                if (entry.hora_salida) {
                    // Esto no debería pasar por la query, pero por seguridad
                    return res.status(400).json({
                        success: false,
                        error: 'Salida ya registrada',
                        message: 'Esta entrada ya tiene salida registrada'
                    });
                }

                // Calcular horas trabajadas
                const horaEntrada = new Date(`${fecha}T${entry.hora_entrada}`);
                const horaSalidaDate = new Date(`${fecha}T${horaSalida}`);
                const horasTrabajadas = ((horaSalidaDate - horaEntrada) / (1000 * 60 * 60)).toFixed(2);

                // Preparar observaciones con información GPS si está disponible
                const observacionesCompletas = [
                    observaciones || 'Salida registrada',
                    req.gpsValidation ? `GPS validado - Distancia: ${req.gpsValidation.distance}m` : null
                ].filter(Boolean).join(' | ');

                // Actualizar registro con salida y coordenadas GPS si están disponibles
                const updateQuery = `
                    UPDATE asistencia 
                    SET hora_salida = ?, 
                        observaciones = COALESCE(observaciones || ' | ', '') || ?,
                        latitud_salida = ?,
                        longitud_salida = ?
                    WHERE id = ?
                `;

                db.run(
                    updateQuery,
                    [horaSalida, observacionesCompletas, latitud || null, longitud || null, entry.id],
                    function(err) {
                        if (err) {
                            console.error('Error registrando salida:', err);
                            return res.status(500).json({
                                success: false,
                                error: 'Error registrando salida'
                            });
                        }

                        res.json({
                            success: true,
                            message: 'Salida registrada correctamente',
                            data: {
                                id: entry.id,
                                fecha,
                                hora_entrada: entry.hora_entrada,
                                hora_salida: horaSalida,
                                horas_trabajadas: parseFloat(horasTrabajadas),
                                usuario: req.user.nombre,
                                gps_validation: req.gpsValidation ? {
                                    distance: req.gpsValidation.distance,
                                    max_distance: req.gpsValidation.maxDistance,
                                    location_valid: req.gpsValidation.isValid
                                } : null
                            }
                        });
                    }
                );
            }
        );

    } catch (error) {
        console.error('Error en fichado salida:', error);
        res.status(500).json({
            success: false,
            error: 'Error interno del servidor'
        });
    }
});

// GET /api/fichado/historial - Obtener historial de fichados con información GPS
router.get('/historial', authenticateToken, (req, res) => {
    try {
        const usuarioId = req.user.id;
        const { limite = 30, pagina = 1, fecha_inicio, fecha_fin } = req.query;
        const offset = (pagina - 1) * limite;

        // Construir filtros de fecha si se proporcionan
        let whereClause = 'WHERE usuario_id = ?';
        let queryParams = [usuarioId];

        if (fecha_inicio) {
            whereClause += ' AND fecha >= ?';
            queryParams.push(fecha_inicio);
        }

        if (fecha_fin) {
            whereClause += ' AND fecha <= ?';
            queryParams.push(fecha_fin);
        }

        const query = `
            SELECT 
                a.id,
                a.fecha,
                a.hora_entrada,
                a.hora_salida,
                CASE 
                    WHEN a.hora_entrada IS NOT NULL AND a.hora_salida IS NOT NULL 
                    THEN ROUND((julianday(a.fecha || ' ' || a.hora_salida) - julianday(a.fecha || ' ' || a.hora_entrada)) * 24, 2)
                    ELSE NULL 
                END as horas_trabajadas,
                a.latitud,
                a.longitud,
                a.latitud_salida,
                a.longitud_salida,
                a.metodo_fichado,
                a.observaciones,
                a.timestamp_creacion,
                CASE 
                    WHEN a.latitud IS NOT NULL AND a.longitud IS NOT NULL 
                    THEN 'GPS_DISPONIBLE'
                    ELSE 'SIN_GPS'
                END as estado_gps,
                u.nombre,
                u.apellido,
                u.cargo,
                u.area,
                u.email
            FROM asistencia a
            INNER JOIN usuarios u ON a.usuario_id = u.id
            ${whereClause.replace('WHERE usuario_id', 'WHERE a.usuario_id')}
            ORDER BY a.fecha DESC, a.hora_entrada DESC
            LIMIT ? OFFSET ?
        `;

        queryParams.push(parseInt(limite), offset);

        db.all(query, queryParams, (err, rows) => {
            if (err) {
                console.error('Error obteniendo historial:', err);
                return res.status(500).json({
                    success: false,
                    error: 'Error obteniendo historial'
                });
            }

            // Obtener total de registros para paginación
            const countQuery = `SELECT COUNT(*) as total FROM asistencia a ${whereClause.replace('WHERE usuario_id', 'WHERE a.usuario_id')}`;
            const countParams = queryParams.slice(0, -2); // Remover limite y offset

            db.get(countQuery, countParams, (err, countResult) => {
                if (err) {
                    console.error('Error contando registros:', err);
                    return res.status(500).json({
                        success: false,
                        error: 'Error obteniendo historial'
                    });
                }

                // Calcular estadísticas del período
                const totalHoras = rows.reduce((sum, row) => sum + (row.horas_trabajadas || 0), 0);
                const diasTrabajados = rows.filter(row => row.hora_entrada && row.hora_salida).length;
                const registrosIncompletos = rows.filter(row => row.hora_entrada && !row.hora_salida).length;

                res.json({
                    success: true,
                    data: rows,
                    pagination: {
                        total: countResult.total,
                        pagina: parseInt(pagina),
                        limite: parseInt(limite),
                        total_paginas: Math.ceil(countResult.total / limite)
                    },
                    estadisticas: {
                        total_horas_trabajadas: Math.round(totalHoras * 100) / 100,
                        dias_trabajados: diasTrabajados,
                        registros_incompletos: registrosIncompletos,
                        promedio_horas_dia: diasTrabajados > 0 ? Math.round((totalHoras / diasTrabajados) * 100) / 100 : 0
                    }
                });
            });
        });

    } catch (error) {
        console.error('Error en historial:', error);
        res.status(500).json({
            success: false,
            error: 'Error interno del servidor'
        });
    }
});

// GET /api/fichado/estado-hoy - Estado del fichado de hoy
router.get('/estado-hoy', authenticateToken, (req, res) => {
    try {
        const usuarioId = req.user.id;
        const fecha = new Date().toISOString().split('T')[0];

        db.get(
            `SELECT 
                id,
                fecha,
                hora_entrada,
                hora_salida,
                CASE 
                    WHEN hora_entrada IS NOT NULL AND hora_salida IS NOT NULL 
                    THEN ROUND((julianday(fecha || ' ' || hora_salida) - julianday(fecha || ' ' || hora_entrada)) * 24, 2)
                    ELSE NULL 
                END as horas_trabajadas,
                metodo_fichado,
                observaciones
            FROM asistencia 
            WHERE usuario_id = ? AND fecha = ?`,
            [usuarioId, fecha],
            (err, row) => {
                if (err) {
                    console.error('Error obteniendo estado de hoy:', err);
                    return res.status(500).json({
                        success: false,
                        error: 'Error obteniendo estado'
                    });
                }

                const estado = {
                    fecha,
                    tiene_entrada: false,
                    tiene_salida: false,
                    puede_marcar_entrada: true,
                    puede_marcar_salida: false,
                    horas_trabajadas: 0,
                    registro: null
                };

                if (row) {
                    estado.tiene_entrada = !!row.hora_entrada;
                    estado.tiene_salida = !!row.hora_salida;
                    estado.puede_marcar_entrada = !row.hora_entrada;
                    estado.puede_marcar_salida = !!row.hora_entrada && !row.hora_salida;
                    estado.horas_trabajadas = row.horas_trabajadas || 0;
                    estado.registro = row;
                }

                res.json({
                    success: true,
                    data: estado
                });
            }
        );

    } catch (error) {
        console.error('Error en estado-hoy:', error);
        res.status(500).json({
            success: false,
            error: 'Error interno del servidor'
        });
    }
});

module.exports = router;